File: src/bin/day6.rs
use itertools::Itertools;
use std::collections::HashMap;
fn main() {
    let input = include_str!("../../input/day6.txt");
    let part1: usize = input
        .split("\n\n")
        .map(|l| l.chars().filter(|c| *c != '\n').unique().count())
        .sum();
    println!("{}", part1);

    let part2: usize = input
        .split("\n\n")
        .map(|l| {
            let mut counts: HashMap<char, usize> = HashMap::new();
            let people = l.chars().filter(|c| *c == '\n').count() + 1;
            l.chars()
                .filter(|c| *c != '\n')
                .for_each(|c| *counts.entry(c).or_insert(0) += 1);
            counts.values().filter(|val| **val == people).count()
        })
        .sum();
    println!("{}", part2);
}

File: python/day6.py
from itertools import combinations
import math

with open("input/day6.txt", "r") as f:
    read = f.read().split("\n\n")
    groups = [[set(i) for i in group.split()] for group in read]
    # Part 1
    total = sum(map(lambda sets: len(set.union(*sets)), groups))
    print(total)
    # Part 2
    total = sum(map(lambda sets: len(set.intersection(*sets)), groups))
    print(total)